### PR TITLE
Final clean-ups in <input>

### PR DESCRIPTION
**Pull request by TimothyGu**
This is rebased on top of #2786. Creating a draft for Travis to take a look at it first.

### CODE REVIEW

File: lib/jsdom/living/helpers/strings.js [hunk 1]
```diff
@@ -1,8 +1,13 @@
 "use strict";
 
+// https://infra.spec.whatwg.org/#ascii-whitespace
 const asciiWhitespaceRe = /^[\t\n\f\r ]$/;
 exports.asciiWhitespaceRe = asciiWhitespaceRe;
 
+// https://infra.spec.whatwg.org/#ascii-digit
+const asciiDigitRe = /^[0-9]$/;
+exports.asciiDigitRe = asciiDigitRe;
```
Comment (domenic) [inline lib/jsdom/living/helpers/strings.js]: This appears to be unused?

File: lib/jsdom/living/nodes/HTMLProgressElement-impl.js [hunk 2]
```diff
@@ -10,38 +10,58 @@ class HTMLProgressElementImpl extends HTMLElementImpl {
...
+        return parsedMax;
+      }
     }
-
     return 1.0;
   }
   set max(value) {
-    this.setAttributeNS(null, "max", value);
+    if (value > 0) {
+      this.setAttributeNS(null, "max", String(value));
```
Comment (domenic) [inline lib/jsdom/living/nodes/HTMLProgressElement-impl.js]: The string conversion is not necessary due to https://github.com/jsdom/jsdom/commit/4285e7074367c56dda07d3c08d6528ff0651cd4d#diff-802e42090f68edfaab3d0f57d32de76bR249

File: lib/jsdom/living/nodes/HTMLInputElement-impl.js [hunk 3]
```diff
@@ -144,27 +144,22 @@ class HTMLInputElementImpl extends HTMLElementImpl {
...
-      if (v > candidate) {
-        candidate += this._allowedValueStep;
-      }
-    } else if (v < candidate) {
-      candidate -= this._allowedValueStep;
-    }
-    return candidate + this._stepBase;
+    const allowedValueStep = this._allowedValueStep;
+    const stepBase = this._stepBase;
+    return new Decimal(v).minus(stepBase)
```
Comment (domenic) [inline lib/jsdom/living/nodes/HTMLInputElement-impl.js]: I think these changes should be tested.

File: lib/jsdom/living/nodes/HTMLInputElement-impl.js [hunk 4]
```diff
@@ -967,7 +953,8 @@ class HTMLInputElementImpl extends HTMLElementImpl {
       // than the minimum, the element is simultaneously suffering from an underflow and suffering
       // from an overflow.
       const reversedRangeSufferingOverUnderflow = () => {
-        return this._parsedValueIsNumber && this._parsedValue > this._maximum && this._parsedValue < this._minimum;
+        const parsedValue = this._convertStringToNumber(this._value);
+        return typeof parsedValue === "number" && parsedValue > this._maximum && parsedValue < this._minimum;
```
Comment (domenic) [inline lib/jsdom/living/nodes/HTMLInputElement-impl.js]: Can we replace the typeof === number with !== null?

File: lib/jsdom/living/helpers/strings.js (hunk 1 above)
Comment (TimothyGu) [reply lib/jsdom/living/helpers/strings.js]: Indeed. Removed.

File: lib/jsdom/living/nodes/HTMLInputElement-impl.js (hunk 4 above)
Comment (TimothyGu) [reply lib/jsdom/living/nodes/HTMLInputElement-impl.js]: Done.

File: lib/jsdom/living/nodes/HTMLProgressElement-impl.js (hunk 2 above)
Comment (TimothyGu) [reply lib/jsdom/living/nodes/HTMLProgressElement-impl.js]: Sure. I'd want to eventually remove the type cast though. Maybe after reflection gets re-done.

Comment (domenic) [reply lib/jsdom/living/nodes/HTMLProgressElement-impl.js]: I wouldn't be comfortable doing that unless we had some sort of static enforcement (e.g. lint rule) or auto-generated testing infrastructure to catch future errors... it's just too easy to forget that the useful behavior from the wrapper is omitted in the impl...

**Comment by domenic** [comment]
The tests are failing here, and I still think that we need tests for the decimal changes.

**Comment by TimothyGu** [comment]
I spent quite a bit of time trying to make a test case for the decimal change but so far I haven’t been able to. I did uncover some discrepancies between Firefox and Chrome with regards to precision though, even they are both using a big decimal implementation. (The same test has the results in jsdom before/after the change.) The tricky part is that the spec doesn’t seem to mandate a specific level of precision, so it’s unclear who exactly is correct.

I’ll fix the test failure when I get to it.

**Comment by TimothyGu** [comment]
In light of https://github.com/whatwg/html/issues/5207, I'd rather not add tests for the decimal change. We already know that it "works" in the sense that it doesn't regress what used to work. Otherwise, fixed the test failures.

**Comment by domenic** [comment]
But, why include it at all, if it doesn't help us in a testable observable way? What if we regress, with no tests to catch us? I'd rather remove that change, if we don't think it's worth testing.

**Comment by domenic** [comment]
I'm going to pull the other two commits into master :)

**Comment by domenic** [comment]
Oh, I guess that isn't so straightforward, since the third depends on the under-discussion second. Nevermind, for now!

**Comment by TimothyGu** [comment]
Okay, I've removed the controversial change to `_stepAlign` and the commit is now a simple refactoring. See if this works better.

**Pull request merged**